ledger/participant-integration-api/src/main/scala/platform/store/backend/common/ContractStorageBackendTemplate.scala

```diff
@@ -0,0 +1,432 @@
+  )(connection: Connection): Option[StorageBackend.RawContract] = {
+    activeContract(
+      resultSetParser = contractRowParser.singleOpt,
+      resultColumns = List("template_id", "create_argument", "create_argument_compression"),
+    )(
+      readers = readers,
+      contractId = contractId,
+    )(connection)
+  }
+
+  override def activeContractWithoutArgument(
+      readers: Set[Ref.Party],
+      contractId: ContractId,
+  )(connection: Connection): Option[String] =
+    activeContract(
+      resultSetParser = contractWithoutValueRowParser.singleOpt,
+      resultColumns = List("template_id"),
+    )(
+      readers = readers,
+      contractId = contractId,
+    )(connection)
+
+  override def contractKey(readers: Set[Ref.Party], key: Key)(
+      connection: Connection
+  ): Option[ContractId] =
+    contractKey(
+      resultColumns = List("contract_id"),
+      resultParser = contractId("contract_id"),
+    )(
+      readers = Some(readers),
+      key = key,
+      validAt = None,
+    )(connection)
+
+  private def contractKey[T](
+      resultColumns: List[String],
+      resultParser: RowParser[T],
+  )(
+      readers: Option[Set[Ref.Party]],
+      key: Key,
+      validAt: Option[Long],
+  )(
+      connection: Connection
+  ): Option[T] = {
+    def withAndIfNonEmptyReaders(
+        queryF: Set[Ref.Party] => (String, List[NamedParameter])
+    ): (String, List[NamedParameter]) =
+      readers match {
+        case Some(readers) =>
+          val pair = queryF(readers)
+          pair.copy(_1 = s"${pair._1} AND")
+
+        case None =>
+          ("", Nil)
+      }
+
+    val (lastContractKeyFlatEventWitnessesClause, lastContractKeyFlatEventWitnessesParams) =
+      withAndIfNonEmptyReaders(
+        queryStrategy.arrayIntersectionNonEmptyClause(
+          columnName = "last_contract_key_create.flat_event_witnesses",
+          _,
+          paramNamePostfix = "lastContractKey",
+        )
+      )
+    val (participantEventsFlatEventWitnessesClause, participantEventsFlatEventWitnessesParams) =
+      withAndIfNonEmptyReaders(
+        queryStrategy.arrayIntersectionNonEmptyClause(
+          columnName = "participant_events.flat_event_witnesses",
+          _,
+          paramNamePostfix = "participantEvents",
+        )
+      )
+    val (validAtClause, validAtParams) = validAt match {
+      case Some(validAt) =>
+        (
+          "AND event_sequential_id <= {validAt}",
+          List[NamedParameter]("validAt" -> validAt),
+        )
+
+      case None =>
+        ("", Nil)
+    }
+
+    import com.daml.platform.store.Conversions.HashToStatement
+    val sql =
+      s"""  WITH last_contract_key_create AS (
+         |         SELECT participant_events.*
+         |           FROM participant_events, parameters
+         |          WHERE event_kind = 10 -- create
+         |            AND create_key_hash = {keyHash}
+         |            AND event_sequential_id <= parameters.ledger_end_sequential_id
```

Comment:
Do you need this line if you have the `$validAtClause` below? I would have expected this line to be folded into the `validAtClause` if `validAt` is `None`.

Reply:
Probably not. But to do this we also need to remove the join as well. That's why I was thinking:
* it is simpler
* and also safer
To just let the two conditions be merged to one in the SQL engine.
Provided that it is not significantly less efficient which would motivate tradeoff on safety. (so benchmarks need to verify this cc @tudor-da )

Reply:
Makes sense :+1: